File: Lab 9 (17 April 2025)/n-Queens.cpp
#include <iostream>
#include <vector>
using namespace std;

bool isSafe(int row, int col, vector<string>& board, int n)
{
    for (int i = 0; i < row; ++i)
        if (board[i][col] == 'Q') return false;

    for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; --i, --j)
        if (board[i][j] == 'Q') return false;

    for (int i = row - 1, j = col + 1; i >= 0 && j < n; --i, ++j)
        if (board[i][j] == 'Q') return false;

    return true;
}

void solve(int row, vector<string>& board, int n, vector<vector<string>>& solutions)
{
    if (row == n)
    {
        solutions.push_back(board);
        return;
    }

    for (int col = 0; col < n; ++col)
    {
        if (isSafe(row, col, board, n))
        {
            board[row][col] = 'Q';
            solve(row + 1, board, n, solutions);
            board[row][col] = '.';
        }
    }
}

void nQueens(int n)
{
    vector<vector<string>> solutions;
    vector<string> board(n, string(n, '.'));
    
    solve(0, board, n, solutions);

    cout << "Total solutions: " << solutions.size() << endl;
    for (const auto& sol : solutions)
    {
        for (const string& row : sol)
            cout << row << endl;
        cout << endl;
    }
}

int main()
{
    int n;
    cout << "Enter number of queens: ";
    cin >> n;
    nQueens(n);
    return 0;
}

File: Lab 9 (17 April 2025)/n queens.py
import time
import matplotlib.pyplot as plt

def is_safe(board, row, col, n):
    for i in range(col):
        if board[row][i] == 1:
            return False
    for i, j in zip(range(row-1, -1, -1), range(col-1, -1, -1)):
        if board[i][j] == 1:
            return False
    for i, j in zip(range(row+1, n), range(col-1, -1, -1)):
        if board[i][j] == 1:
            return False
    return True

def solve(board, col, n):
    if col >= n:
        return True
    for i in range(n):
        if is_safe(board, i, col, n):
            board[i][col] = 1
            if solve(board, col + 1, n):
                return True
            board[i][col] = 0
    return False

def n_queens(n):
    board = [[0]*n for _ in range(n)]
    return solve(board, 0, n)

def measure_execution_time():
    sizes = list(range(4, 16))  # n from 4 to 15
    times = []
    for n in sizes:
        start = time.time()
        n_queens(n)
        end = time.time()
        times.append(end - start)
    return sizes, times

sizes, times = measure_execution_time()

plt.plot(sizes, times, marker='o', color='purple')
plt.title("Time Complexity of N-Queens (Backtracking)")
plt.xlabel("Number of Queens (n)")
plt.ylabel("Execution Time (seconds)")
plt.grid(True)
plt.show()